app/controllers/training_controller.rb

```diff
@@ -59,6 +60,15 @@ def reload
     render plain: e.message
   end
 
+  def find_slide
+    training_slide = TrainingSlide.find(params[:slide_id])
+    training_module = find_module_from_slide_slug(training_slide.slug)
+    raise ActionController::RoutingError, 'module not found' unless training_module
+    training_library = find_library_from_module_slug(training_module.slug)
+    raise ActionController::RoutingError, 'library not found' unless training_library
```

Comment:
It's possible to have a module that is not in any library, and it would render if you visited a path that had any valid training library slug. (You can view a module from one library even if you substitute the slug from another library.)

It would be good to find the library as you do here, but if no library is found, it would be better to use TrainingLibrary.first as a fallback instead of throwing a 404.

Reply:
hmm interesting. That behavior is not so intuitive I guess.

I changed the logic for the controllers and updated specs on fbb894f8765a8f9a1537ed3ea056bdcdcc714eea

Note that the new definitions may result in failure if no library exists. In such a scenario, calling `TrainingLibrary.first` would return `nil`, causing an error when attempting to access the slug.
Not sure if a world with no libraries makes sense, or it is not worth considering this case and we can suppose `TrainingLibrary.first` is always well defined. 
It looks like modules and slides render even using a non-existing library slug, so we could use a literal string
`'mylibrary'` as default library slug value if no libraries at all.
What's your opinion here?